simulator/object.go

```diff
@@ -0,0 +1,56 @@
+/*
+Copyright (c) 2017-2018 VMware, Inc. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package simulator
+
+import (
+	"github.com/vmware/govmomi/vim25/methods"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/soap"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+func SetCustomValue(e mo.Entity, req *types.SetCustomValue) soap.HasFault {
+	body := &methods.SetCustomValueBody{}
+
+	cfm := Map.CustomFieldsManager()
+
+	_, field := cfm.findByNameType(req.Key, req.This.Type)
+	if field == nil {
+		body.Fault_ = Fault("", &types.InvalidArgument{InvalidProperty: "key"})
+		return body
+	}
+
+	entity := Map.Get(req.This).(mo.Entity).Entity()
+	// Temporarily unlock this object for SetField to use
```

Comment:
I used Unlock there because `custom_fields_manager.go` `SetField(...)` method uses `Map.WithLock(...)`, and that was causing the program to lock up in an infinite wait. I'm not sure how to use the context to prevent that.

Now I'm thinking it might be better to separate `SetCustomValue(...)` (in `object.go`) and `SetField(...)` (in `custom_fields_manager.go`) logic instead of making one call the other - I wanted to reuse the code. In short - one method is called on the entity, and should not be trying to lock it when setting its fields, the other is called on `CustomFieldsManager` and should lock the entity whose fields it's setting.

Reply:
I used Unlock there as `SetField(...)` method was causing the program to lock up in an infinite wait:
https://github.com/vmware/govmomi/blob/af4d22f023a160295d1f9e239bf5985673878587/simulator/custom_fields_manager.go#L103-L109
I'm not sure how to use the context to prevent that.

Now I'm thinking it might be better to separate `SetCustomValue` (in `object.go`) and `SetField` (in `custom_fields_manager.go`), because now one of them calls the other to reuse the code. The problem is that when setting field from the entity itself (`SetCustomValue`), it should not be trying to relock itself, but when field is set from CustomFieldManager (`SetField`), the entity that's being updated should be locked.

What do you think?

Reply:
Sorry, I meant to add, just change `Map.WithLock` to `ctx.WithLock`, here's where it checks Caller to avoid the lock: https://github.com/vmware/govmomi/blob/af4d22f023a160295d1f9e239bf5985673878587/simulator/session_manager.go#L279-L281